File: train/train2.py
#!/usr/bin/python3
"""
train2.py: Bigram language model with a single-layer MLP, trained with autograd.

Same as train1.py:
- Dataset, tokenizer, model architecture (MLP), SGD optimizer, inference

Different from train1.py:
- Introduces autograd (Value class) to compute gradients automatically
- No more manual analytic_gradient or numerical_gradient
- The forward pass builds a computation graph, loss.backward() traverses it

The hand-derived chain rule from train1.py is now automated: each operation
records its local gradient, and backward() applies the chain rule recursively.
The training loop becomes: forward pass -> loss.backward() -> SGD update.
"""

import os       # os.path.exists
import math     # math.log, math.exp
import random   # random.seed, random.choices, random.gauss, random.shuffle
random.seed(42)

# Dataset: load and tokenize a list of names
if not os.path.exists('input.txt'):
    import urllib.request
    names_url = 'https://raw.githubusercontent.com/karpathy/makemore/refs/heads/master/names.txt'
    urllib.request.urlretrieve(names_url, 'input.txt')
docs = [l.strip() for l in open('input.txt').read().strip().split('\n') if l.strip()] # list[str] of documents
random.shuffle(docs)
print(f"num docs: {len(docs)}")

# Tokenizer: character-level, with a special BOS (Beginning of Sequence) token
uchars = sorted(set(''.join(docs))) # unique characters in the dataset become token ids 0..n-1
BOS = len(uchars) # token id for the special Beginning of Sequence (BOS) token
vocab_size = len(uchars) + 1 # total number of unique tokens, +1 is for BOS
print(f"vocab size: {vocab_size}")

# Autograd: recursively apply the chain rule through a computation graph
class Value:
    __slots__ = ('data', 'grad', '_children', '_local_grads')

    def __init__(self, data, children=(), local_grads=()):
        self.data = data
        self.grad = 0
        self._children = children
        self._local_grads = local_grads

    def __add__(self, other):
        other = other if isinstance(other, Value) else Value(other)
        return Value(self.data + other.data, (self, other), (1, 1))

    def __mul__(self, other):
        other = other if isinstance(other, Value) else Value(other)
        return Value(self.data * other.data, (self, other), (other.data, self.data))

    def __pow__(self, other): return Value(self.data**other, (self,), (other * self.data**(other-1),))
    def log(self): return Value(math.log(self.data), (self,), (1/self.data,))
    def exp(self): return Value(math.exp(self.data), (self,), (math.exp(self.data),))
    def relu(self): return Value(max(0, self.data), (self,), (float(self.data > 0),))
    def __neg__(self): return self * -1
    def __radd__(self, other): return self + other
    def __sub__(self, other): return self + (-other)
    def __rsub__(self, other): return other + (-self)
    def __rmul__(self, other): return self * other
    def __truediv__(self, other): return self * other**-1
    def __rtruediv__(self, other): return other * self**-1

    def backward(self):
        topo = []
        visited = set()
        def build_topo(v):
            if v not in visited:
                visited.add(v)
                for child in v._children:
                    build_topo(child)
                topo.append(v)
        build_topo(self)
        self.grad = 1
        for v in reversed(topo):
            for child, local_grad in zip(v._children, v._local_grads):
                child.grad += local_grad * v.grad

# Initialize the parameters
n_embd = 16     # embedding dimension
matrix = lambda nout, nin: [[Value(random.gauss(0, 0.08)) for _ in range(nin)] for _ in range(nout)]
state_dict = {
    'wte': matrix(vocab_size, n_embd),
    'mlp_fc1': matrix(4 * n_embd, n_embd),
    'mlp_fc2': matrix(vocab_size, 4 * n_embd),
}
params = [p for mat in state_dict.values() for row in mat for p in row]
print(f"num params: {len(params)}")

# Model: token_id -> logits
def linear(x, w):
    return [sum(wi * xi for wi, xi in zip(wo, x)) for wo in w]

def softmax(logits):
    max_val = max(val.data for val in logits)
    exps = [(val - max_val).exp() for val in logits]
    total = sum(exps)
    probs = [e / total for e in exps]
    return probs

def mlp(token_id):
    x = state_dict['wte'][token_id]
    x = linear(x, state_dict['mlp_fc1'])
    x = [xi.relu() for xi in x]
    logits = linear(x, state_dict['mlp_fc2'])
    return logits

# Train the model
num_steps = 1000
learning_rate = 1.0
for step in range(num_steps):

    # Take single document, tokenize it, surround it with BOS special token on both sides
    doc = docs[step % len(docs)]
    tokens = [BOS] + [uchars.index(ch) for ch in doc] + [BOS]
    n = len(tokens) - 1

    # Forward pass
    losses = []
    for pos_id in range(n):
        token_id, target_id = tokens[pos_id], tokens[pos_id + 1]
        logits = mlp(token_id)
        probs = softmax(logits)
        loss_t = -probs[target_id].log()
        losses.append(loss_t)
    loss = (1 / n) * sum(losses)

    # Backward pass
    loss.backward()

    # SGD update
    lr_t = learning_rate * (1 - step / num_steps) # linear learning rate decay
    for i, p in enumerate(params):
        p.data -= lr_t * p.grad
        p.grad = 0

    # if step < 5 or step % 200 == 0:
    print(f"step {step+1:4d} / {num_steps:4d} | loss {loss.data:.4f}")

# Inference: sample new names from the model
temperature = 0.5
print("\n--- inference (new, hallucinated names) ---")
for sample_idx in range(20):
    token_id = BOS
    sample = []
    for pos_id in range(16):
        logits = mlp(token_id)
        probs = softmax([l / temperature for l in logits])
        token_id = random.choices(range(vocab_size), weights=[p.data for p in probs])[0]
        if token_id == BOS:
            break
        sample.append(uchars[token_id])
    print(f"sample {sample_idx+1:2d}: {''.join(sample)}")

File: train/train2.ml
(** Bigram language model with a single-layer MLP, trained with autograd. *)

(*
Same as train1.ml:
- Dataset, tokenizer, model architecture (MLP), SGD optimizer, inference

Different from train1.ml:
- Introduces autograd to compute gradients automatically
- No more manual analytic_gradient or numerical_gradient
- The forward pass builds a computation graph, loss.backward() traverses it

The hand-derived chain rule from train1.py is now automated: each operation
records its local gradient, and backward() applies the chain rule recursively.
The training loop becomes: forward pass -> loss.backward() -> SGD update.
 *)

open Extlib
open Autograd

type state =
  {
    wte : Matrix.t;
    mlp_fc1 : Matrix.t;
    mlp_fc2 : Matrix.t;
  }

let () =
  Random.self_init ();

  (* Let there be a Dataset docs: a list of documents (e.g. a list of names). *)
  if not (Sys.file_exists "input.txt") then File.download "https://raw.githubusercontent.com/karpathy/makemore/refs/heads/master/names.txt" "input.txt";
  let docs =
    File.read "input.txt"
    |> String.split_on_char '\n'
    |> List.map String.trim
    |> List.filter (fun s -> s <> "")
    |> Array.of_list
  in
  Array.shuffle docs;

  (* Tokenizer: character-level, with a special BOS (Beginning of Sequence) token *)
  let uchars =
    (* unique characters in the dataset become token ids *)
    Array.to_seq docs
    |> Seq.concat_map String.to_seq
    |> List.of_seq
    |> List.sort_uniq compare
    |> Array.of_list
  in
  (* token id for a special Beginning of Sequence (BOS) token *)
  let bos = Array.length uchars in
  (* total number of unique tokens, +1 is for BOS *)
  let vocab_size = Array.length uchars + 1 in
  Printf.printf "vocab size: %d\n%!" vocab_size;

  (* Initialize the parameters *)
  let n_embd = 16 in (* embedding dimension *)
  let matrix nout nin = Matrix.init nout nin (fun _ _ -> const (0.08 *. Random.gauss ())) in
  let state =
    {
      wte = matrix vocab_size n_embd;
      mlp_fc1 = matrix (4 * n_embd) n_embd;
      mlp_fc2 = matrix vocab_size (4 * n_embd);
    }
  in
  let params = List.flatten @@ List.map Matrix.coefficients [state.wte; state.mlp_fc1; state.mlp_fc2] in
  Printf.printf "num params: %d\n" (List.length params);

  let mlp token_id =
    state.wte.(token_id)
    |> Matrix.ap state.mlp_fc1
    |> Vector.map relu
    |> Matrix.ap state.mlp_fc2
  in

  (* Train the model *)
  let num_steps = 1000 in
  let learning_rate = 1. in
  for step = 0 to num_steps do

    (* Take single document, tokenize it, surround it with BOS special token on both sides *)
    let tokens =
      docs.(step mod Array.length docs)
      |> String.to_seq
      |> Seq.map (Array.index uchars)
      |> (fun doc -> Seq.concat (List.to_seq [Seq.return bos; doc; Seq.return bos]))
      |> Array.of_seq
    in
    let n = Array.length tokens - 1 in

    (* Forward pass *)
    let loss = ref @@ const 0. in
    for pos_id = 0 to n - 1 do
      let token_id = tokens.(pos_id) in
      let target_id = tokens.(pos_id + 1) in
      let logits = mlp token_id in
      let probs = Vector.soft_max logits in
      let loss_t = neg (log probs.(target_id)) in
      loss := add !loss loss_t
    done;
    let loss = cmul (1. /. float n) !loss in

    (* Backward pass *)
    backward loss;

    (* SGD update *)
    let lr_t = learning_rate *. (1. -. float step /. float num_steps) in (* linear learning rate decay *)
    List.iter (fun p ->
        p.value <- value p -. lr_t *. grad p;
        p.grad <- 0.
      ) params;

    if step < 5 || step mod 100 = 0 then
    Printf.printf "step %4d / %4d | loss %.4f\n%!" step num_steps (value loss)
  done;

  (* Inference: sample new names from the model *)
  print_endline "--- inference (new, hallucinated names) ---";
  let temperature = 0.5 in
  let block_size = 16 in (* maximum sequence length *)
  for sample_idx = 0 to 20 - 1 do
    let token_id = ref bos in
    let sample = ref [] in
    let pos_id = ref 0 in
    while !pos_id < block_size do
      incr pos_id;
      let logits = mlp !token_id in
      let probs = Vector.soft_max @@ Vector.cmul (const (1. /. temperature)) logits in
      token_id := Random.index @@ Array.map value @@ probs;
      if !token_id = bos then pos_id := block_size
      else sample := uchars.(!token_id) :: !sample;
    done;
    let sample = String.of_seq @@ List.to_seq @@ List.rev !sample in
    Printf.printf "sample %2d: %s\n%!" (sample_idx+1) sample
  done
